libgstd/libgstd.c

```diff
@@ -0,0 +1,198 @@
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following
+ * disclaimer in the documentation and/or other materials provided
+ * with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#define _GNU_SOURCE
+#include <stdarg.h>
+#include <stdio.h>
+
+#include "libgstd.h"
+#include "gstd_ipc.h"
+#include "gstd_tcp.h"
+#include "gstd_unix.h"
+#include "gstd_http.h"
+#include "gstd_log.h"
+#include "libgstd_assert.h"
+
+static GType gstd_supported_ipc_to_ipc (SupportedIpcs code);
+static void gstd_manager_init (GOptionGroup ** gst_group, int argc,
+    char *argv[]);
+
+struct _GstDManager
+{
+  GstdSession *session;
+  GstdIpc **ipc_array;
+  guint num_ipcs;
+};
+
+static GType
+gstd_supported_ipc_to_ipc (SupportedIpcs code)
+{
+  GType code_description[] = {
+    [GSTD_IPC_TYPE_TCP] = GSTD_TYPE_TCP,
+    [GSTD_IPC_TYPE_UNIX] = GSTD_TYPE_UNIX,
+    [GSTD_IPC_TYPE_HTTP] = GSTD_TYPE_HTTP
+  };
+
+  const gint size = sizeof (code_description) / sizeof (gchar *);
+
+  gstd_assert_and_ret_val (0 <= code, GSTD_TYPE_IPC);   // TODO: Proponer un GSTD_TYPE_DEFAULT
+  gstd_assert_and_ret_val (size > code, GSTD_TYPE_IPC);
+
+  return code_description[code];
+}
+
+static void
+gstd_manager_init (GOptionGroup ** gst_group, int argc, char *argv[])
+{
+  gst_init (&argc, &argv);
+  gstd_debug_init ();
+
+  if (gst_group != NULL && *gst_group != NULL) {
+    *gst_group = gst_init_get_option_group ();
+  }
+
+}
+
+GstdStatus
+gstd_manager_new (SupportedIpcs supported_ipcs[], guint num_ipcs,
+    GstDManager ** out, GOptionGroup ** gst_group, int argc, char *argv[])
+{
+  GstDManager *manager;
+  GstdSession *session;
+  GstdStatus ret = GSTD_LIB_OK;
+  GstdIpc **ipc_array;
+
+  gstd_assert_and_ret_val (NULL != out, GSTD_NULL_ARGUMENT);
+
+  manager = (GstDManager *) malloc (sizeof (GstDManager));
+  session = gstd_session_new ("Session0");
+
+  /* If there is ipcs, then initialize them */
+  if (NULL != supported_ipcs) {
+    ipc_array = g_malloc (num_ipcs * sizeof (GstdIpc *));
```

Comment:
`
    ipc_array = g_malloc0 (num_ipcs * sizeof (*ipc_array));
`

Reply:
What happens if `num_ipcs` is 0?

Reply:
Validation added